### DISCUSSION TITLE
Git and Premake - Automate premake to run after branch switch/checkout with a git hook!

### DISCUSSION THREAD

**Discussion by danieldownes** [premake/premake-core#2221] [Show and tell]
Hi there,

I started on helping out a friend on a C++ project game VS project. He just gone and saved his VS files directly in the repo of course, like anyone else might. We had some junior devs help out, and they had issues setting up the project, yes, as the project requirement some manual configuration to the project on a per machine basis of course Premake was a clear way forward.

However, as we have to checkout branches all the time to check each others work, and the fact that the project is in flux and has a lot of changes, people tend to forget to run Premake. Most VS devs don't exactly use the command line tbh, that's why we are using a heavy IDE. This would include the git functionalities in the IDE!

My idea was to automate running premake each time the branch is switched, in git terms this is a *checkout*. There is a way to hook into git events using *git hooks*. Git hooks are not versioned or shared by default when you commit changes to a repository. 

Of course I wanted to share. This is not something people do on a daily basis, especially junior devs. I added a bat file to the repository to add the git hook so that every time a git checkout is successfully performed, then permake is run.

I created a setup-hooks.bat file. which is in the root of the repo:

```
@echo off
REM Setup script to copy hooks to the .git/hooks directory and set permissions using Git Bash

REM Get the path to the root of the repository
for /f "delims=" %%i in ('git rev-parse --show-toplevel') do set REPO_ROOT=%%i

REM Copy the post-checkout hook file
copy "%REPO_ROOT%\.githooks\post-checkout" "%REPO_ROOT%\.git\hooks\post-checkout"

REM Check if Git Bash exists
set GIT_BASH_PATH=C:\Program Files\Git\bin\bash.exe
if exist "%GIT_BASH_PATH%" (
    REM Use Git Bash to set the executable permission
    "%GIT_BASH_PATH%" -c "chmod +x \"%REPO_ROOT%/.git/hooks/post-checkout\""
    echo Git hooks have been set up.
) else (
    echo Warning: Git Bash not found. Please install Git for Windows from https://gitforwindows.org and rerun this script.
)

pause
```

This post-checkout file looks like this:
```
#!/bin/sh
# Hook script to run a batch file after checking out a branch

# Change to the root of the repository
cd "$(git rev-parse --show-toplevel)"

# Path to your batch file
BATCH_FILE="premake.bat"

# Run the batch file
cmd //c $BATCH_FILE
```

Finally the premake.bat file calls premake:

```
start vendor\premake5.exe vs2022
```
We use a bat file so that it can be manually run more easily, and also we add some of other project prep here, like copying in dlls to the bin folders, etc.

PS, the game is here, using SFML only and lots of code!
https://github.com/TheEmperorsArmoury/TheGreatWar/

**Comment by Jarod42** [comment]
Notice there is now (Since Premake 5.0.0 beta 3) [`gitintegration`](https://premake.github.io/docs/gitintegration/)
